apps/prairielearn/src/ee/auth/lti13/lti13Auth.ts

```diff
@@ -149,9 +158,58 @@ const OIDCAuthResponseSchema = z.object({
 router.post(
   '/callback',
   asyncHandler(async (req, res) => {
+    debug('callback');
+    debug(req.method);
+    debug(req.body);
+
+    // https://www.imsglobal.org/spec/security/v1p0/#step-3-authentication-response
+    OIDCAuthResponseSchema.passthrough().parse(req.body);
+
     const lti13_instance = await selectLti13Instance(req.params.lti13_instance_id);
 
-    const lti13_claims = await authenticate(req, res);
+    const openidClientConfig = await getOpenidClientConfig(lti13_instance);
+
+    // Needed for implicit flow
+    client.useIdTokenResponseType(openidClientConfig);
+
+    // URL href doesn't matter, openid-client only uses the url.hash to pass properties
+    // into client.implicitAuthentication
+    const url = new URL('https://example.com/');
+    url.hash = new URLSearchParams(req.body).toString();
+
+    const IDToken = await client.implicitAuthentication(
+      openidClientConfig,
+      url,
+      req.session.lti13_state.nonce,
+      {
+        expectedState: req.session.lti13_state.state,
+      },
+    );
+
+    const lti13_claims = Lti13ClaimSchema.parse(IDToken);
+
+    // Check nonce to protect against reuse
+    const nonceKey = `lti13auth-nonce:${req.params.lti13_instance_id}:${lti13_claims['nonce']}`;
+    const cacheResult = await cache.get(nonceKey);
+    if (cacheResult) {
+      throw new HttpStatusError(500, 'Cannot reuse LTI 1.3 nonce, try login again');
+    }
+    cache.set(nonceKey, true, 60 * 60 * 1000); // 60 minutes
+    // Canvas OIDC logins expire after 3600 seconds
+
+    // Remove auth state from session
+    delete req.session.lti13_state;
+
+    // Save parameters about the platform back to the lti13_instance
+    // https://www.imsglobal.org/spec/lti/v1p3#platform-instance-claim
+    const params = {
+      lti13_instance_id: req.params.lti13_instance_id,
+      tool_platform_name:
+        lti13_claims['https://purl.imsglobal.org/spec/lti/claim/tool_platform']?.name ?? null,
+    };
```

Comment:
Nit: no separate params objects, just inline this into the second argument of `queryAsync`.